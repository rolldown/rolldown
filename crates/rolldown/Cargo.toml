[package]
edition.workspace    = true
homepage.workspace   = true
license.workspace    = true
name                 = "rolldown"
repository.workspace = true
version              = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lints]
workspace = true

[dependencies]
anyhow             = { workspace = true }
dunce              = { workspace = true }
futures            = { workspace = true }
indexmap           = { workspace = true }
itertools          = { workspace = true }
once_cell          = { workspace = true }
oxc                = { workspace = true }
oxc_index          = { workspace = true }
regex              = { workspace = true }
rolldown_common    = { workspace = true }
rolldown_error     = { workspace = true }
rolldown_fs        = { workspace = true, features = ["os"] }
rolldown_lang_json = { workspace = true }
rolldown_oxc_utils = { workspace = true }
rolldown_plugin    = { workspace = true }
rolldown_resolver  = { workspace = true }
rolldown_rstr      = { workspace = true }
rolldown_sourcemap = { workspace = true }
rolldown_tracing   = { workspace = true }
rolldown_utils     = { workspace = true }
rustc-hash         = { workspace = true }
sugar_path         = { workspace = true }
tokio              = { workspace = true, features = ["rt", "macros", "sync"] }
tracing            = { workspace = true }
tracing-chrome     = { workspace = true }
xxhash-rust        = { workspace = true, features = ["xxh3"] }

[dev-dependencies]
glob             = { workspace = true }
insta            = { workspace = true }
rolldown_testing = { workspace = true }
sugar_path       = { workspace = true }
testing_macros   = { workspace = true }
tokio            = { workspace = true, features = ["rt", "macros", "sync", "rt-multi-thread"] }
