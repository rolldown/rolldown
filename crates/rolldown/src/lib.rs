mod ast_scanner;
mod bundler;
mod bundler_builder;
mod chunk;
mod chunk_graph;
mod error;
mod finalizer;
mod module_loader;
mod runtime;
mod stages;
mod types;
mod utils;

use std::sync::Arc;

use rolldown_fs::OsFileSystem;
use rolldown_resolver::Resolver;

pub(crate) type SharedResolver = Arc<Resolver<OsFileSystem>>;
pub(crate) type SharedOptions = Arc<NormalizedBundlerOptions>;

pub use crate::{
  bundler::Bundler, bundler_builder::BundlerBuilder, chunk::render_chunk::PreRenderedChunk,
  types::rolldown_output::RolldownOutput,
};

pub use rolldown_common::bundler_options::*;

pub use rolldown_resolver::ResolveOptions;
