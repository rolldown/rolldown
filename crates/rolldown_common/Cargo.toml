[package]
description = "This crate is mostly for sharing code between rolldwon crates."
name        = "rolldown_common"
version     = "0.0.1"

edition.workspace    = true
homepage.workspace   = true
license.workspace    = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lints]
workspace = true

[dependencies]
anyhow             = { workspace = true }
index_vec          = { workspace = true }
once_cell          = { workspace = true }
oxc                = { workspace = true, features = ["semantic"] }
regex              = { workspace = true }
rolldown_fs        = { workspace = true }
rolldown_rstr      = { workspace = true }
rolldown_sourcemap = { workspace = true }
rolldown_utils     = { workspace = true }
rustc-hash         = { workspace = true }
schemars           = { workspace = true }
serde              = { workspace = true, features = ["derive"] }
sugar_path         = { workspace = true }

[features]
deserialize_bundler_options = []
