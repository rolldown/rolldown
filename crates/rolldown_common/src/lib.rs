mod chunk;
mod inner_bundler_options;
mod module;
mod types;

/// This module is to help `rolldown` crate could export types related bundler options easily.
/// `rolldown` crate could use `pub use rolldown_common::bundler_options::*;` to export all types, so we don't need write
/// the same code in `rolldown` crate again.
pub mod bundler_options {
  pub use crate::inner_bundler_options::{
    types::{
      filename_template::{FileNameRenderOptions, FilenameTemplate},
      input_item::InputItem,
      is_external::IsExternal,
      loader::Loader,
      normalized_bundler_options::NormalizedBundlerOptions,
      output_format::OutputFormat,
      output_option::{AddonFunction, AddonOutputOption},
      platform::Platform,
      resolve_options::ResolveOptions,
      source_map_type::SourceMapType,
      sourcemap_ignore_list::SourceMapIgnoreList,
      sourcemap_path_transform::SourceMapPathTransform,
    },
    BundlerOptions,
  };
}

// We don't want internal position adjustment of files affect users, so all items are exported in the root.
pub use crate::{
  chunk::{
    types::{
      cross_chunk_import_item::CrossChunkImportItem, pre_renderer_chunk::PreRenderedChunk,
      preliminary_filename::PreliminaryFilename,
    },
    Chunk,
  },
  module::external_module::ExternalModule,
  module::normal_module::NormalModule,
  types::ast_scope::AstScope,
  types::bundler_file_system::BundlerFileSystem,
  types::chunk_id::ChunkId,
  types::chunk_kind::ChunkKind,
  types::entry_point::{EntryPoint, EntryPointKind},
  types::exports_kind::ExportsKind,
  types::external_module_id::ExternalModuleId,
  types::import_record::{ImportKind, ImportRecord, ImportRecordId, RawImportRecord},
  types::importer_record::ImporterRecord,
  types::module_id::ModuleId,
  types::module_info::ModuleInfo,
  types::module_table::{ExternalModuleVec, ModuleTable, NormalModuleVec},
  types::module_type::ModuleType,
  types::named_export::LocalExport,
  types::named_import::{NamedImport, Specifier},
  types::normal_module_id::NormalModuleId,
  types::output::{Output, OutputAsset},
  types::output_chunk::OutputChunk,
  types::package_json::PackageJson,
  types::rendered_chunk::RenderedChunk,
  types::rendered_module::RenderedModule,
  types::resolved_export::ResolvedExport,
  types::resolved_path::ResolvedPath,
  types::resolved_request_info::ResolvedRequestInfo,
  types::resource_id::{representative_name, ResourceId},
  types::side_effects,
  types::stmt_info::{DebugStmtInfoForTreeShaking, StmtInfo, StmtInfoId, StmtInfos},
  types::symbol_ref::SymbolRef,
  types::wrap_kind::WrapKind,
};
pub use bundler_options::*;
