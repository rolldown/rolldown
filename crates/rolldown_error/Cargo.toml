[package]
description = "rolldown_error"
name        = "rolldown_error"
version     = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

edition.workspace    = true
homepage.workspace   = true
license.workspace    = true
repository.workspace = true

[lints]
workspace = true

[dependencies]
anyhow            = { workspace = true }
arcstr            = { workspace = true }
ariadne           = { workspace = true }
napi              = { workspace = true, optional = true }
oxc               = { workspace = true }
rolldown_resolver = { workspace = true }
sugar_path        = { workspace = true }
