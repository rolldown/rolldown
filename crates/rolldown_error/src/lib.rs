mod build_error;
mod diagnostic;
mod event_kind;
mod events;
mod utils;

use std::{borrow::Cow, path::Path};

use sugar_path::SugarPath;

pub use crate::{
  build_error::{BuildError, BuildResult},
  event_kind::EventKind,
};

trait PathExt {
  fn relative_display(&self) -> Cow<str>;
}

impl PathExt for Path {
  fn relative_display(&self) -> Cow<str> {
    let cwd = std::env::current_dir().expect("Failed to get current directory");
    let non_absolute = if self.is_absolute() {
      Cow::Owned(self.relative(cwd).to_slash_lossy().into_owned())
    } else {
      self.to_string_lossy()
    };

    non_absolute
  }
}
