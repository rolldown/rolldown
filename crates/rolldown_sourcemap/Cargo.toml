[package]
name    = "rolldown_sourcemap"
version = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

edition.workspace    = true
homepage.workspace   = true
license.workspace    = true
repository.workspace = true

[lints]
workspace = true

[lib]
bench   = false
doctest = false
test    = false

[dependencies]
memchr = { workspace = true }
oxc    = { workspace = true }
