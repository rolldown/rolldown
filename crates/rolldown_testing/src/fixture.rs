use std::path::{Path, PathBuf};

use crate::{integration_test::IntegrationTest, test_config::read_test_config};

use rolldown_testing_config::TestConfig;

pub struct Fixture {
  config_path: PathBuf,
  fixture_path: PathBuf,
}

impl Fixture {
  pub fn new(path: impl AsRef<Path>) -> Self {
    // Paths could be UNC format in windows, see https://github.com/rust-lang/rust/issues/42869 for more details
    let path = dunce::simplified(path.as_ref());

    Self { fixture_path: path.to_path_buf(), config_path: path.join("_config.json") }
  }

  pub fn run_integration_test(self) {
    tokio::runtime::Runtime::new().unwrap().block_on(self.run_inner());
  }

  async fn run_inner(self) {
    let TestConfig { config: mut options, meta } = read_test_config(&self.config_path);

    if options.cwd.is_none() {
      options.cwd = Some(self.fixture_path.clone());
    }

    IntegrationTest::new(meta).run(options).await;
  }
}
