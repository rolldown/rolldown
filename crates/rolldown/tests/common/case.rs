use std::{borrow::Cow, path::Path};

use super::fixture::Fixture;
use rolldown::RolldownOutput;
use rolldown_common::Output;
use rolldown_error::BuildError;
use rolldown_sourcemap::SourcemapVisualizer;

pub struct Case {
  fixture: Fixture,
  snapshot: String,
}

impl Case {
  pub fn new(path: impl AsRef<Path>) -> Self {
    // Paths could be UNC format in windows, see https://github.com/rust-lang/rust/issues/42869 for more details
    let path = dunce::simplified(path.as_ref());

    Self { fixture: Fixture::new(path.to_path_buf()), snapshot: String::new() }
  }

  pub fn run(self) {
    std::env::set_var("ROLLDOWN_TEST", "1");
    tokio::runtime::Runtime::new().unwrap().block_on(self.run_inner());
  }

  pub async fn run_inner(mut self) {
    let build_output = self.fixture.compile().await;
    match build_output {
      Ok(assets) => {
        assert!(!self.fixture.test_config().expect_error, "expected error, but got success");
        self.render_assets_to_snapshot(assets);
      }
      Err(errs) => {
        assert!(
          self.fixture.test_config().expect_error,
          "expected success, but got errors: {errs:?}"
        );
        self.render_errors_to_snapshot(errs);
      }
    }
    self.make_snapshot();
    self.fixture.exec();
  }

  fn render_assets_to_snapshot(&mut self, outputs: RolldownOutput) {
    let mut assets = outputs.assets;
    let warnings = outputs.warnings;

    if !warnings.is_empty() {
      self.snapshot.push_str("# warnings\n\n");
      let diagnostics = warnings.into_iter().map(|e| (e.code(), e.into_diagnostic()));
      let rendered = diagnostics
        .flat_map(|(code, diagnostic)| {
          [
            Cow::Owned(format!("## {code}\n")),
            "```text".into(),
            Cow::Owned(diagnostic.to_string()),
            "```".into(),
          ]
        })
        .collect::<Vec<_>>()
        .join("\n");
      self.snapshot.push_str(&rendered);
      self.snapshot.push('\n');
    }

    self.snapshot.push_str("# Assets\n\n");
    assets.sort_by_key(|c| c.file_name().to_string());
    let artifacts = assets
      .iter()
      .filter(|asset| !asset.file_name().contains("$runtime$") && matches!(asset, Output::Chunk(_)))
      .flat_map(|asset| {
        [
          Cow::Owned(format!("## {}\n", asset.file_name())),
          "```js".into(),
          Cow::Borrowed(asset.content().trim()),
          "```".into(),
        ]
      })
      .collect::<Vec<_>>()
      .join("\n");
    self.snapshot.push_str(&artifacts);

    if self.fixture.test_config().snapshot_output_stats {
      self.render_stats_to_snapshot(&assets);
    }

    if self.fixture.test_config().visualize_sourcemap {
      self.render_sourcemap_visualizer_to_snapshot(&assets);
    }
  }

  fn render_stats_to_snapshot(&mut self, assets: &[Output]) {
    self.snapshot.push_str("\n\n## Output Stats\n\n");
    let stats = assets
      .iter()
      .flat_map(|asset| match asset {
        Output::Chunk(chunk) => {
          vec![Cow::Owned(format!(
            "- {}, is_entry {}, is_dynamic_entry {}, exports {:?}",
            chunk.file_name, chunk.is_entry, chunk.is_dynamic_entry, chunk.exports
          ))]
        }
        Output::Asset(_) => vec![],
      })
      .collect::<Vec<_>>()
      .join("\n");
    self.snapshot.push_str(&stats);
  }

  fn render_errors_to_snapshot(&mut self, mut errors: Vec<BuildError>) {
    self.snapshot.push_str("# Errors\n\n");
    errors.sort_by_key(BuildError::code);
    let diagnostics = errors.into_iter().map(|e| (e.code(), e.into_diagnostic()));

    let rendered = diagnostics
      .flat_map(|(code, diagnostic)| {
        [
          Cow::Owned(format!("## {code}\n")),
          "```text".into(),
          Cow::Owned(diagnostic.to_string()),
          "```".into(),
        ]
      })
      .collect::<Vec<_>>()
      .join("\n");
    self.snapshot.push_str(&rendered);
  }

  fn render_sourcemap_visualizer_to_snapshot(&mut self, assets: &[Output]) {
    self.snapshot.push_str("\n\n# Sourcemap Visualizer\n\n");
    let visualizer_result = assets
      .iter()
      .filter_map(|asset| match asset {
        Output::Chunk(chunk) => chunk
          .map
          .as_ref()
          .map(|sourcemap| SourcemapVisualizer::new(&chunk.code, sourcemap).into_visualizer_text()),
        Output::Asset(_) => None,
      })
      .collect::<Vec<_>>()
      .join("\n");
    self.snapshot.push_str(&visualizer_result);
  }

  fn make_snapshot(&self) {
    // Configure insta to use the fixture path as the snapshot path
    let fixture_folder = self.fixture.dir_path();
    let mut settings = insta::Settings::clone_current();
    let content = self.snapshot.to_string();
    settings.set_snapshot_path(fixture_folder);
    settings.set_prepend_module_to_snapshot(false);
    settings.set_input_file(fixture_folder);
    settings.bind(|| {
      insta::assert_snapshot!("artifacts", content);
    });
  }
}
