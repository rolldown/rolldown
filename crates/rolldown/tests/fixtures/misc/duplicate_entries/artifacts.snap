---
source: crates/rolldown/tests/common/case.rs
expression: content
input_file: crates/rolldown/tests/fixtures/misc/duplicate_entries
---
# Assets

## main.mjs

```js
import { main_default } from "./main~1.mjs";

export { main_default as default };
```
## main2.mjs

```js
import { main_default } from "./main~1.mjs";

export { main_default as default };
```
## main~1.mjs

```js
// main.js
var main_default = {};

export { main_default };
```
